src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -549,25 +548,37 @@ public void close()
                         delegate.close();
                     }
 
-                    @Override
-                    public boolean hasNext()
+                    private Row computeNext()
                     {
-                        while (hasNext = delegate.hasNext())
+                        while (delegate.hasNext())
                         {
-                            next = delegate.next();
+                            Row row = delegate.next();
                             queryContext.rowsFiltered++;
-                            if (tree.isSatisfiedBy(delegate.partitionKey(), next, staticRow))
-                                return true;
+                            if (tree.isSatisfiedBy(delegate.partitionKey(), row, staticRow))
+                                return row;
                         }
-                        return false;
+                        return null;
+                    }
+
+                    @Override
+                    public boolean hasNext()
+                    {
+                        if (next == null)
+                            next = computeNext();
+                        return next != null;
                     }
 
                     @Override
                     public Row next()
                     {
-                        if (!hasNext)
+                        if (next == null)
+                            next = computeNext();
+                        if (next == null)
```

Comment:
These 3 rows can be replaced with `if (!hasNext())`.

Reply:
Right, technically they could be, but I don't agree they should be. 
Replacing that with `hasNext()` would make the code rely on a side-effect that's not a part of the `hasNext()` contract, but only an internal implementation detail.